Processing pipeline for historical installers

sofiapinto opened the pull request:
# Description 

This PR:
- creates a processing pipeline for historical MCS installers data;
- creates getters to retrieve raw historical installers and installations data from S3 (I've added the data manually to S3, to `asf-core-data/inputs/MCS/latest_raw_data/historical/` - had to create a folder for historical data within `latest_raw_data`, otherwise it would have messed the work in `generate_mcs_data.py` because of the file names);
- adds the company unique ID to the installations data in the processing installations pipeline;
- creates tests for the processing pipeline;
- I've also added a notebook (`asf_core_data/analysis/historical_installers.ipynb`) if you want to take a look at raw data and processed data, while reviewing the code, to better understand the changes.

closes #31 
closes #45

# Instructions for Reviewer(s)

Hey @ch-williamson, @athrado and @sqr00t thanks so much for taking the time to look at this PR.
- Chris and Julia, it would be great if at least one of you could check the files run, but more than that, check if all makes sense within the asf_core_data data processing pipeline and that the additions/changes don't have potential for messing something else up :) as a follow-up we should probably add a few tests that are automatically run by github actions.
- Solomon, it would be great if you could check if the general "flow" looks good to you/makes sense and if any changes are needed before the code gets transferred to asf_daps. Note that, the historical installers data needs to have a pair of variables as table primary key: `(company_unique_id, mcs_certificate_number)`
- Julia, I didn't add code to create a merged version of the data (for HPMT consumption) that includes the installers, as I didn't have time to familiarise enough with `mcs_epc_joining.py` and `merge_install_dates.py` scripts, to integrate these changes somewhere in the code. But hopefully it shouldn't be too difficult. All that needs to be done is an outer merge between processed installers and installations. Happy to do that after I'm back, or for you to do it if you would prefer.

## Setup
- Create a developer account for Companies House API (it's super quick to set up, but the confirmation email might take a few minutes to arrive, maybe about 10-15min):
   1. Register for an account [here](https://developer.company-information.service.gov.uk/get-started);
   2. Create a new application [here](https://developer.company-information.service.gov.uk/manage-applications/add). Give it any name and description you want and choose the Live environment.
   3.  Copy your API key credentials.
- clone this repo: `git clone git@github.com:nestauk/asf_core_data.git`
- checkout to the correct branch: `git checkout 31_historical_MCS_installers_data`
- Run `make install`;
- Run `direnv allow`;
- Activate the conda enviroment: `conda activate asf_core_data`;
- Set your API credentials as environment variable, i.e run `export COMPANIES_HOUSE_API_KEY="ADD_YOUR_API_KEY_HERE"` in the command line and replace `ADD_YOUR_API_KEY_HERE` with your API key credentials.

## Review
Could you run the following script to check if it works for you?
`python3 asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py`
A file will be saved to S3 and replace the one I have previously created (absolutely no problem with that, feel free to run!). It gives a warning, `SettingWithCopyWarning`, but all should run fine and a file should be saved to S3: `asf-core-data/outputs/MCS/installers/mcs_historical_installers_20230207.csv`

There's a script to test the functions I created. Could you please check if 17 out 18 tests pass? Should take approximately 2 mins to run.
`pytest asf_core_data/pipeline/mcs/test/test_process_historical_mcs_installers.py`
One of the tests fails, the one testing the data coming from Companies House - it doesn't mean that the function is not doing what's supposed to do. It just means that 1) the accuracy of the matching is low or 2) the matching is good but the data is incorrect in one of the sources (when comparing postcode data coming from MCS to Companies House data). I've manually checked some of the values and it seems that sometimes Companies House has the correct postcode, other times the correct postcode is in MCS data. But I need to take a further look at this!
Also please feel free to suggest other tests.

I would also ask you to review the remaining scripts with changes, with a bigger focus on: 
- `asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py`
- The additions made to `asf_core_data/pipeline/mcs/process/process_mcs_installations.py` - is this a good place to add the "company_unique_id" var to the installations? Or maybe it should be added to `generate_mcs_data.py`?

A few questions:
- Shall we keep the postcode with the space or without?
- Does the general pre-processing pipeline for historical installers make sense? Any suggestions for simplification/to improve readability?

Thanks a lot!

---

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [ ] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [ ] I have explained the feature in this PR or (better) in `output/reports/`
- [x] I have requested a code review

sofiapinto committed:
processing pipeline for historical installers

sofiapinto committed:
changes to processing pipeline, adding getters and config variables

sofiapinto committed:
adding company unique id to installations data

sofiapinto committed:
adding script with functions to test historical installers processing pipeline

sofiapinto committed:
adding notebook to compare raw and processed data

athrado commented on asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py:
```diff
@@ -0,0 +1,926 @@
+        raw_historical_installations.columns
+    )
+
+    raw_historical_installations["certification_body"] = (
+        raw_historical_installations["installation_company_mcs_number"]
+        .str.split(" ")
+        .str[0]
+    )
+
+    raw_historical_installations["installation_company_mcs_number"] = (
+        raw_historical_installations["installation_company_mcs_number"]
+        .str.split(" ")
+        .str[1]
+        .astype(int)
+    )
```
All good. In case you want to save a few lines, you could do this in one step. Look at dummy example below. You'd still have to modify the company_mcs_number with `.astype(int)`, though.

![Screenshot 2023-02-25 at 15 02 17](https://user-images.githubusercontent.com/42718928/221364321-41f4ac21-76bb-4c1a-b532-02026646b9c8.png)

sofiapinto replied:
very cool suggestion, didn't know about `expand`. Will change, thanks so much! :D

athrado commented on asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py:
```diff
@@ -0,0 +1,926 @@
+        Renamed column names
+    """
+    cols = [c.lower() for c in cols]
+    changes_dict = {"heat pump": "hp", " ": "_", "/": "_", "adddress": "address"}
```
Didn't check whether this ends up being a problem in practice, but since dicts are not ordered, I worry that the following could happen: In the string 'heat pump', the white space is replaced with '_' and so 'heat_pump' will not be shortened to 'hp'. Easiest fix is adding an additonal replacement entry:

```suggestion
    changes_dict = {"heat pump": "hp", "heat_pump":"hp", " ": "_", "/": "_", "adddress": "address"}
```

sofiapinto replied:
I don't think this would ever be a problem in practice, but I will add a test to check this in the test script :)

athrado commented on asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py:
```diff
@@ -0,0 +1,926 @@
+    for col in installation_cols:
+        new_date_var = col.replace("installation_", "")
+        correspondent_design_var_name = col.replace("installation", "design")
+        data[new_date_var] = data.apply(
+            lambda x: x[col]
+            if not pd.isnull(x[col])
+            else x[correspondent_design_var_name],
+            axis=1,
```
Looks fine and I assume you tested that this does exactly what it should, but I haven't double-checked myself.

sofiapinto replied:
Yup, I have a test to check the output from this function in `test_process_mcs_historical_installers.py`, the function `test_join_installation_and_design_vars()` - but actually think I just found a bug in the test!

athrado commented on asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py:
```diff
@@ -0,0 +1,926 @@
+        "a trading name of": "trading as",
+        "the trading name of": "trading as",
+    }
+    for item in versions:
+        installer_name = installer_name.replace(item, versions[item])
```
Looks good! Would it be worth adding a `.strip()` at the end of the installer name to get rid of any trailing while spaces, or does that never happen?

athrado replied:
Looking at the code further down, I guess that's not a problem.

sofiapinto replied:
Nop, I think the `clean_company_name()` function takes care of that :)
I can add a further test to the `test_process_historical_mcs_installers.py` script, but all trailing white spaces are dealt with in that function.

Example:
```
company_name = "\tabc t/a  b"
company_name = clean_company_name(company_name)
```
The value of company_name is 'abc ta b'

athrado commented on asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py:
```diff
@@ -0,0 +1,926 @@
+
+
+# copied from process_mcs_utils.py and changed slightly
+def match_companies_house(company_name: str, api_key: str) -> pd.Series:
+    """
```
Nicely done and well-explained!

sofiapinto replied:
thank you :)

athrado commented on asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py:
```diff
@@ -0,0 +1,926 @@
+            and ("items" in company_data.keys())
+            and len(company_data["items"]) > 0
+        ):
+            # return data for the closest match, the one found in index 0
+            if ("address_snippet" in company_data["items"][0].keys()) and (
+                "date_of_creation" in company_data["items"][0].keys()
+            ):
+                return pd.Series(
+                    [
+                        company_data["items"][0]["address_snippet"],
+                        company_data["items"][0]["date_of_creation"],
+                    ]
+                )
+            elif "address_snippet" in company_data["items"][0].keys():
+                return pd.Series(
+                    [
+                        company_data["items"][0]["address_snippet"],
+                        None,
+                    ]
+                )
+            elif "date_of_creation" in company_data["items"][0].keys():
+                return pd.Series(
+                    [
+                        None,
+                        company_data["items"][0]["date_of_creation"],
+                    ]
+                )
+            else:
+                return pd.Series([None, None])
```
If I understand correctly, this could be handled with less code/make it a bit more reabable. Maybe something like:

```suggestion
           address_snippet = company_data["items"][0]["address_snippet"] if ("address_snippet" in company_data["items"][0].keys()) else None
           date_of_creation = company_data["items"][0]["date_of_creation"] if  "date_of_creation" in company_data["items"][0].keys() else None
           
           return  pd.Series(
                    [address_snippet, date_of_creation])
        
```

sofiapinto replied:
definitely! thank you :D

athrado commented on asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py:
```diff
@@ -0,0 +1,926 @@
+            return pd.Series([None, None])
+    else:
+        if response_status_code >= 400 and response_status_code < 500:
+            raise Exception(
+                "Cannot get data, the program will stop!\nHTTP {}: {}".format(
+                    response_status_code, response.text
+                )
+            )
+        sleep_seconds = random.randint(5, 60)
+        print(
+            "Cannot get data, your program will sleep for {} seconds...\nHTTP {}: {}".format(
+                sleep_seconds, response_status_code, response.text
```
Very nice way of handling this! It actually happened to me so I know it's working!

sofiapinto replied:
great! :D

athrado commented on asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py:
```diff
@@ -0,0 +1,926 @@
+    """
+    if pd.isnull(date_creation) or pd.isnull(first_commisioning_date):
+        return full_address
+    return full_address if date_creation <= first_commisioning_date else None
```
Makes sense!

sofiapinto replied:
cool :) really glad

athrado commented on asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py:
```diff
@@ -0,0 +1,926 @@
+        ["date_of_creation"],
+        axis=1,
+        inplace=True,
+    )
```
For .rename and .drop functions, you can also use they keyword argument 'columns' instead of defining the axis. Doens't really matter but I'm always paranoid about mixing up the axes. :D 

```suggestion
    missing_installers.rename(columns=
        {
            "installation_company_name": "company_name",
            "installation_company_mcs_number": "mcs_certificate_number",
        }
        inplace=True,
    )

    # dropping unecessary variables
    missing_installers.drop(columns=
        ["date_of_creation"],
        inplace=True,
    )
```

sofiapinto replied:
yep, using `columns` or `axis=1` is the same, but i'll change it :) it's definitely easier to understand with `columns` :)

athrado commented on asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py:
```diff
@@ -0,0 +1,926 @@
+    Returns:
+        full adress
+    """
+    if original_record:
+        temp = address_1
+        if not pd.isnull(address_2):
+            temp = temp + address_2 + ", "
+        if not pd.isnull(town):
+            temp = temp + town + ", "
+        if not pd.isnull(county):
+            temp = temp + county + ", "
+        if not pd.isnull(postcode):
+            temp = temp + postcode + ", "
+        return temp[:-2]
```
That looks fine, but you could also try something like to prevent repetition and confusion about `temp[:-2]`

```suggestion
    if original_record:
        temp = []
        for address_part in [address_1, address_2, town, county, postcode]:
              if not pd.isnull(address_part):
                    temp.append(address_part)
        full_address = ", ".join(temp)
        return full_address
 
```

sofiapinto replied:
very neat! will add :)

athrado commented on asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py:
```diff
@@ -0,0 +1,926 @@
+    Updates historical installers data by creating flags representing whether
+    installer is certified or not for a certain technology.
+
+    If a start date exists, then if means installer is certified.
```
```suggestion
    If a start date exists, then it means installer is certified.
```

athrado commented on asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py:
```diff
@@ -0,0 +1,926 @@
+        comp for comp in company_name.split() if comp not in company_stopwords
+    ]
+
+    return " ".join(company_name)
```
I assume it's fine that the result is lowercase.

sofiapinto replied:
yep, it's fine, because we don't do these changes to the original company name. We apply them to get the company unique id :)

athrado commented on asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py:
```diff
@@ -0,0 +1,926 @@
+    dictionary = dict()
+    for value in list_values[1:]:
+        dictionary[value] = list_values[0]
+    return dictionary
```
Why would we need this...? I'll probably find out later but it seems odd.  

Either way, you can do this much simpler (see also dummy example). Hardly worth having it's own function. 

![Screenshot 2023-02-25 at 17 04 23](https://user-images.githubusercontent.com/42718928/221369980-8db8ead6-1d46-411c-8100-e1561f28511c.png)

```suggestion
def from_list_to_dictionary(list_values):
    """
    Transforms a list into a dictionary where each key is matched to the first
    value in the list. Creates a mapping between each item in the list and the first one.
    Arg:
        list_values: a list of values e.g. ["Company A", "Company B", "Company A trading as Company B"]
    Returns:
        Dictionary mapping all values to the first one
        e.g. {"Company B":"Company A", "Company A trading as Company B": "Company A"}
    """
    return {value : list_values[0] for value in list_values[1:]}
```

athrado commented on asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py:
```diff
@@ -0,0 +1,926 @@
+        A dictionary mapping company names.
+    """
+
+    # For each instance in data, it creates a ist with all possible versions of each company names
```
```suggestion
    # For each instance in data, it creates a list with all possible versions of each company names
```

athrado commented on asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py:
```diff
@@ -0,0 +1,926 @@
+
+    # For each instance in data, it creates a ist with all possible versions of each company names
+    # splitting string by "trading as". E.g. "A trading as B" -> ["A", "B"]
+    trading_as_companies["list"] = trading_as_companies[
```
I think `list` and `dictionary` are not suitable column names. More specific?

sofiapinto replied:
Yup, I can change those to `list_of_companies` and `dictionary_of_companies`, respectively.

athrado commented on asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py:
```diff
@@ -0,0 +1,926 @@
+    # ["A", "B", "A trading as B"] -> {"B":"A", "A trading as B":"A"}
+    trading_as_companies["dictionary"] = trading_as_companies["list"].apply(
+        from_list_to_dictionary
+    )
```
Is there an advantage of saving the dicts in the column 'dictionary' instead of just concatenating all dicts right away...? I may also misunderstand what's happening here but this could potentially be simplified.

sofiapinto replied:
Saving to the column dictionary is just an intermediate step, but happy to do it all in one go. Any suggestions? @athrado

athrado commented on asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py:
```diff
@@ -0,0 +1,926 @@
+
+
+# copied from process_mcs_utils.py and changed slightly
+def geocode_postcode(data: pd.DataFrame, geodata: pd.DataFrame) -> pd.DataFrame:
```
If we don't need the coordinates for this script itself, I would recommend we add the postcode (and all the other geographical data) at the very end after merging (EPC, MCS installations, MCS installers) to keep it simple. Otherwise we'll have code for extracting the long/lats from post in every script.

sofiapinto replied:
I'm adding the coordinates as this was also previously done by India in the work she did on the installers snapshot data, and it might be useful for a few analyses. Just trying to make this as broad as possible, to account for other analyses besides HPMT. :)

athrado commented on asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py:
```diff
@@ -0,0 +1,926 @@
+    # Joining installation and design variables
+    join_installation_and_design_vars(raw_historical_installers)
+
+    # flag "original_records" (if originally in installers or inputed from installations)
+    raw_historical_installers["original_record"] = [
+        True for i in range(len(raw_historical_installers))
+    ]
```
I assume that works, but make sure the order is correct. Are the first aways the original records?

sofiapinto replied:
At this point in the code we only have original records. Only after this, lines 806-809, we add the missing installers, i.e.  non-original records.

athrado commented on asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py:
```diff
@@ -0,0 +1,926 @@
+    ]
+
+
+def create_argparser():
```
This is something that should probably go to utils

sofiapinto replied:
A few other functions should definitely go to a utils file. But I don't think this one should go, as it is only going to be used in this file and can't be used anywhere else. Am I seeing wrong? :) @athrado

athrado replied:
True, maybe once the rest is more modularised this can stay here. It just struck me as 'out-of-character' for this script

athrado replied:
Doc string missing?

sofiapinto replied:
docstring added now :)

athrado commented on asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py:
```diff
@@ -0,0 +1,926 @@
+    parser = create_argparser()
+    args = parser.parse_args()
+
+    # Loading geodata
+    uk_geo_path = base_config.POSTCODE_TO_COORD_PATH
+    uk_geo_data = load_s3_data(s3_bucket_name, uk_geo_path)
```
Again, this might be premature. I would not worry about the geodata yet.

athrado reviewed (commented):
Hey Sofia,

_For now I've only checked the script `process_historical_mcs_installers.py` - but I'll have another go at the rest later._ 

It's super impressive how you handle all the processing steps and clean up and enhance the data. I can tell that a lot of work went into the details of the code and you documented everything nicely. It's clear that you know exactly what you're doing and why it needs to be done. Well done!

I haven't found any major issues with the code, but I have left a few suggestions for how to improve the code, make it more readable or simplify it. Maybe some of it is helpful, but feel free to discard if it doesn't seem suitable to you.

If there is time, I would recommend splitting up this script into multiple scripts with one main pipeline script that uses code from other modules. I think there are a few functions that logically form a block, so they could go together into one module. I think that would make the script more digestible. (I'm myself guilty of having very long scripts, so I totally understand how easily this happens.) 

And while I'm fine with it, I think that pandas-critical people would be very skeptical of all the pandas operations in your code, especially the complex `.apply()` with lambda expressions. I also don't have a better/fast/cleaner solution for that at hand and it's well-documented so fine by me, but I felt I should point it out anyways... Maybe we can learn about alternatives together at some point. ;-) 

I ran `python3 asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py` and it ran through without errors after fixing the `certification_body` colum at the end (see comment in the code review).  The pytest script also yielded the expected 17 out of 18. 

_I'll have another look at the rest next week but from what I've seen there's not much to correct anymore._

sqr00t commented on asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py:
```diff
@@ -0,0 +1,926 @@
+        ]
+    ]
+
+    # Merge installer with installation data to add the certification_body to the installer data
+    installer_data = installer_data.merge(
+        unique_pairs_in_installations,
+        how="left",
+        right_on=installations_match_vars,
+        left_on=installers_match_vars,
+    )
+
```
When this function is called, the input installers and installations dataframes already have certification_body. Merge results in certification_body with _x and _y suffixes. Suggestion to assign _left and _right suffixes respectively for clarity, then merge _right data into _left if there are nulls in _left, using combine_first.

```
    # Merge installer with installation data to add the certification_body to the installer data
    installer_data = installer_data.merge(
        unique_pairs_in_installations,
        how="left",
        left_on=installers_match_vars,
        right_on=installations_match_vars,
        suffixes=('_left','_right')
    )

    # Unify _left and _right certification_body, filling any left nulls with right values
    installer_data["certification_body"] = \
        installer_data["certification_body_left"].combine_first(
            installer_data["certification_body_right"]
        )

sofiapinto replied:
upsi! So sorry about this @sqr00t, I must have somehow (?) changed the location of the function call before adding the last commit. How weird, because all was running fine when I last run it.

I've changed the function call back to where it's supposed to be so that this doesn't happen. 

In short:
- The installers table doesn't originally have the certification body, that's why we do the merge to get the certification body from the installations table;
- However, when we get the missing installers data (from the installations table), we also get the certification body. So the getting the certification body for the original records should be done before we get the missing installers data.

sqr00t commented on asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py:
```diff
@@ -0,0 +1,926 @@
+    create_installer_unique_id(raw_historical_installers)
+
+    # Certification body information
+    raw_historical_installers = add_certification_body_info(
+        raw_historical_installers, raw_historical_installations
```
Line 793 adds 'certification_body' to installations dataframe. Lines 812 to 818 adds 'certification_body' to installers dataframe. Resulting in duplicate 'certification_body' in merge. Please see suggestion referred in this function's definition (Lines 752 to 759) to account for this.

sofiapinto replied:
see my comment above :)

sqr00t reviewed (changes requested):
Suggestion for duplicate certification_body resulting from add_certification_body_info function

sofiapinto committed:
update code to get output from companies house

Co-authored-by: Julia Suter <42718928+athrado@users.noreply.github.com>

sofiapinto committed:
update function to get full address to avoid repetition

Co-authored-by: Julia Suter <42718928+athrado@users.noreply.github.com>

sofiapinto committed:
correct typo in docstring

Co-authored-by: Julia Suter <42718928+athrado@users.noreply.github.com>

sofiapinto committed:
correct typo

Co-authored-by: Julia Suter <42718928+athrado@users.noreply.github.com>

sofiapinto committed:
using columns instead of axis=1 to improve readability

Co-authored-by: Julia Suter <42718928+athrado@users.noreply.github.com>

sofiapinto committed:
simplifying from_list_to_dictionary function

Co-authored-by: Julia Suter <42718928+athrado@users.noreply.github.com>

sofiapinto committed:
missing comma after adding julia's suggestions

athrado commented on asf_core_data/pipeline/mcs/process/process_mcs_installations.py:
```diff
@@ -117,7 +120,22 @@ def identify_clusters(hps, time_interval=base_config.MCS_CLUSTER_TIME_INTERVAL):
+    """
+    Updates installations table by adding the unique installer ID.
+    Args:
+        installations: installations table
+        installers: historical installers table
```
maybe good to mark what dtype is expected, either in doc string or type hint

sofiapinto replied:
good shout, completely missed it! :D

athrado commented on asf_core_data/pipeline/mcs/process/process_mcs_installations.py:
```diff
@@ -133,11 +151,16 @@ def get_processed_installations_data(refresh=True):
-    # )
+
+    # Adding variable with unique installer ID
+    get_installer_unique_id(installations_data, historical_installers_processed_data)
```
`get_installer_unique_id()` doesn't return anything...? 
Where does the data with added unique ID go? Sorry if I missed something.

sofiapinto replied:
changed :)

athrado commented on asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py:
```diff
@@ -0,0 +1,905 @@
+        return match_companies_house(company_name, api_key)
+
+
+def recompute_full_adress(date_creation, first_commisioning_date, full_address):
```
Type hint or doc string

sofiapinto replied:
thanks for the shout :)

athrado commented on asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py:
```diff
@@ -0,0 +1,905 @@
+    return " ".join(company_name)
+
+
+def from_list_to_dictionary(list_values):
```
type hint for consistency

athrado commented on asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py:
```diff
@@ -0,0 +1,905 @@
+    }
+
+
+def map_installers_same_location_different_id(postcodes, installer_data):
```
type hint for consistency

athrado reviewed (commented):
Just a few more comments, mostly regarding consistency in doc strings and type hinting

sofiapinto committed:
implementing remaining Julia's suggestions and sorting out certification body prob pointed out by Solomon

sofiapinto committed:
adding a few missing type hints and docstrings

sofiapinto committed:
changes to process_mcs_installations.py as per julia's suggestions

sofiapinto committed:
script to test the company unique id is being added correctly to the installations table

sofiapinto committed:
improve docstrings and adding a few more tests to test_process_historical_mcs_installers.py

sofiapinto committed:
adding a few functions to the process_mcs_utils.py file

sofiapinto committed:
update code to consider most recent version of historical installers

sofiapinto committed:
update generate_mcs_data.py with code to process and save historical installers

sofiapinto committed:
changes to use most recent version of historical installers and installations from S3

sofiapinto committed:
small improvements to process_historical_mcs_installers.py

sofiapinto committed:
mark process_mcs_installers.py as legacy

sofiapinto committed:
mark a few functions as legacy and move them to end of file

sofiapinto committed:
changes to use historical installations in the installations processing pipeline instead of concatenating quarter files

sofiapinto committed:
sort encoding issue

sofiapinto committed:
adds a function that gets most recent batch name of an s3 folder

sofiapinto committed:
ads further tests to test_process_mcs_installations.py script

sofiapinto committed:
adding a note to the initial description at the top of the file

sofiapinto committed:
remove print in get_most_recent_batch_name() function

sofiapinto committed:
change in extract_token_set to account for missing postcodes and moved a few erroneously tagged legacy functions back up

sofiapinto committed:
changes mcs processing pipeline to use historical installations instead of concatenating quarter files and tags a few functions as legacy

sofiapinto committed:
changes to README.md file

sofiapinto commented:
Hey team @athrado and @sqr00t ,
Thank you so much for your reviews and helpful comments!

Since you've reviewed I have:
1. added Julia's suggestions
3. improved the docstrings in the test file + added additional tests to answer a few of Julia's questions
4. added a new test file for check the outputs of processing installations
5. I've left the code to get lat/long values as is as it might be necessary for a few analysis that only use installers (not for HPMT)
6. sorted the problem with certification body -> in a different way from what you suggested @sqr00t because the function call was basically in the wrong place. Should be sorted now.
7. updated the code to get the latest version of historical installers when processing all MCS data (we currently only have one version, but will be useful when new data comes);
8. added the processing of installers to the `process_and_save_mcs()` function in `generate_mcs_data.py`;
9. while doing 6) I noticed that there were a few problems with using historical data for installers and the quarter data for installations (slight differences between historical installations and quarter installations and the fact that the first few quarters didn't have MCS certification number as a variable, so the values were appearing as missing, when in fact they are not missing). Hence, I have created a new issue #45 that tackles changing the processing of installations to use the historical installations dataset instead of concatenating the quarter installation files.
10. Changed the main README.md to account for the changes.
11. A few small issues came up and I also sorted those: encoding issues, missing postcodes, etc.
12. I've tagged a few functions as legacy and moved them as further down as possible in `process_mcs_utils.py` and `generate_mcs_data.py` as they were previously used but are no longer used, but might be useful not to erase them in case we ever want to run an older analysis and we need these older functions;
13. I've tagged the `process_mcs_installers.py` file as legacy, as we are now working with historical installers data. Didn't delete it because again, we might want to re-run it for an older piece of analysis. I think this file can be excluded from asf-daps though.

@sqr00t I know we've talked about renaming "historical" to "latest" or something similar, but I didn't have time to do it so maybe that can be sorted on asf-daps side.

@sqr00t @athrado, I would ask at least one of you to:
- Run `python3 asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py` and make sure this time it runs without any trouble;
- Run `pytest asf_core_data/pipeline/mcs/test/test_process_historical_mcs_installers.py` - again 17 out 18 tests should pass;
- Run `pytest asf_core_data/pipeline/mcs/test/test_process_mcs_installations.py` - 4/5 tests should pass (1 with warning). one of them doesn't because the full/historical installations file doesn't come with the RHI variables (we can discuss on Monday if we need them and I'll ask Adam to add them back up). It's good to have a test to check for the variables, since now whenever we received new data we can run it to see if the variables come with the same names and if all of them come.
- Run `python3 asf_core_data/pipeline/mcs/generate_mcs_data.py` to make sure it also runs for you. Should take around 30 minutes to run. You need to:
   - have the following folder structure `inputs/EPC/raw_data/2022_Q3_complete` under `asf_core_data`
   - [download this file from S3](https://s3.console.aws.amazon.com/s3/object/asf-core-data?region=eu-west-2&prefix=outputs/EPC/preprocessed_data/2022_Q3_complete/EPC_GB_preprocessed.csv.zip) and place it under `outputs/EPC/preprocessed_data/` and unzipping it. The changes Julia is doing in the other branch will make our life much easier in the future!
- Do let me know if any of the changes look odd, but I don't think you need to look in detail to any changes :) if you do want to do it, you can do so by selecting the changes since you've last reviewed the PR.

Thanks so much for taking the time!

sqr00t commented:
Thank you for the detailed updates and review instructions :)

1. process_historical_mcs_installers.py -> ran successfully (with the known warning)
2. test_process_historical_mcs_installers.py -> 17/18 passed (test_match_companies_house fails)
3. test_process_mcs_installations -> 4/5 passed (test_raw_columns_exist fails)
4. generate_mcs_data.py -> ran successfully, although your last request in the previous comment has a small typo. The downloaded zip file needs to be under outputs/EPC/preprocessed_data/2022_Q3_complete/

w.r.t. "historical" to "latest" tbh, I was actually thinking the same, happy to sort it out in asf_daps :)

sofiapinto committed:
Merge branch 'dev' into 31_historical_MCS_installers_data
